examples/step-78/step-78.cc

```diff
@@ -0,0 +1,869 @@
+    Tensor<1, dim> return_value;
+    for (int i = 0; dim; i++)
+      {
+        return_value += -2 * (this->get_time()) * std::sin(p(i)) * Point<1>(1);
+      }
+    return return_value;
+  }
+
+  /*
+  The next piece is the declaration of the main class of this program. This is
+  very similar to the Step-26 tutorial, with some modifications. New matrices
+  had to be added to calculate the A and B matrices, as well as the $V_{diff}$
+  vector mentioned in the introduction.
+  */
+  template <int dim>
+  class BlackScholes
+  {
+  public:
+    BlackScholes();
+
+    // enum RefinementMode
+    // {
+    //     global_refinement,
+    //     adaptive_refinement
+    // };
+
+    void run();
+
+    const double s_max;
+    const double maturity_time;
+    const double sigma;
+    const double r;
+    const double strike_price;
+
+    // RefinementMode refinement_mode;
+
+  private:
+    void setup_system();
+    void solve_time_step();
+    void solve_time_step_diffusion();
+    void refine_mesh(const unsigned int min_grid_level,
+                     const unsigned int max_grid_level);
+    void refine_grid();
+    void o_results();
+    void process_solution(const unsigned int cycle);
+
+    Triangulation<dim> triangulation;
+    FE_Q<dim>          fe;
+    DoFHandler<dim>    dof_handler;
+
+    AffineConstraints<double> constraints;
+
+    SparsityPattern      sparsity_pattern;
+    SparseMatrix<double> mass_matrix;
+    SparseMatrix<double> laplace_matrix;
+    SparseMatrix<double> a_matrix;
+    SparseMatrix<double> b_matrix;
+    SparseMatrix<double> system_matrix;
+
+    Vector<double> solution;
+    Vector<double> old_solution;
+    Vector<double> system_rhs;
+
+
+    double       time;
+    double       time_step;
+    unsigned int timestep_number;
+
+    const double       theta;
+    const unsigned int n_cycles;
+
+    DataOutStack<dim>        data_out_stack;
+    std::vector<std::string> solution_names;
+
+    ConvergenceTable convergence_table;
+  };
+
+  /*
+  In the following classes and functions, we implement the various pieces of
+  data that define this problem. This includes the boundary values and the
+  initial conditions that are used in this program and for which we need
+  function objects. The initial conditions and boundary values as described in
+  the introduction.
+  */
+
+  // Initial Condition Function
+  template <int dim>
+  class InitialConditions : public Function<dim>
+  {
+  public:
+    InitialConditions(double _s_price);
+    double         s_price;
+    virtual double value(const Point<dim> & p,
+                         const unsigned int component = 0) const override;
+  };
+
+  template <int dim>
+  double InitialConditions<dim>::value(const Point<dim> & p,
+                                       const unsigned int component) const
+  {
+    (void)component;
+    return -std::pow(p(component), 2) + 6;
+    // Below is the original initial condition
+    // return std::max(p(component) - s_price, 0.);
```

Comment:
Could we introduce a parameter somewhere so that we can switch between original version and MMS?